### PR TITLE
fix(v2): Add hooks to detect window resize, toggle off sidebar and navbar in desktop

### LINKED ISSUE #2750: Mobile nav does not close when expanding width
## To Reproduce

(Write your steps here:)

1. Reduce size of screen to mobile size (<~1000px)
2. Open mobile nav (bottom right button)
3. Expand screen to desktop size.
4. The nav is not closing, this is what it looks like on the docusaurus official site

Screen Shot 2020-05-15 at 5 57 45 AM

### PR DESCRIPTION
**Pull request by guillaumejacquart**
## Motivation
Resolve #2750 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

1. Change browser size to mobile (<=996px)
2. Open one of the nav bar or doc sidebar using the mobile toggling button
3. Resize to desktop width (> 996px)
4. Observe reset of mobile navigation

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const desktopSize = 'desktop';
+const mobileSize = 'mobile';
+
+function useWindowSize() {
+  const isClient = typeof window === 'object';
+
+  function getSize() {
+    return isClient
+      ? (window.innerWidth > desktopThresholdWidth && desktopSize) || mobileSize
+      : undefined;
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js:20]: I'd suggest extracting the fn outside of the hook + making it more "readable". 

Usage of `&&` and  `||` is hard to follow


```
function getSize() {
  if (isClient) {
    return undefined;
  }
  returrn window.innerWidth > desktopThresholdWidth ? desktopSize : mobileSize;
}
```

File: packages/docusaurus-theme-classic/src/theme/Navbar/index.js [hunk 2]
```diff
@@ -201,6 +202,14 @@ function Navbar() {
     [setLightTheme, setDarkTheme],
   );
 
+  const windowSize = useWindowSize();
+
+  useEffect(() => {
+    if (windowSize === desktopSize && sidebarShown) {
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Navbar/index.js:208]: in the same way `setWindowSize(getSize());` does not trigger useless re-renders, it's not necessary to test `&& sidebarShown` because React will already noop if `setSidebarShown(false);` is called while `sidebarShown=false`.

(also you didn't capture `sidebarShown` in the deps array)

Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/Navbar/index.js:208]: https://twitter.com/sebastienlorber/status/1272516868994224128

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js (hunk 1 above)
Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js]: Indeed, much clearer. I updated the code.

File: packages/docusaurus-theme-classic/src/theme/Navbar/index.js (hunk 2 above)
Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/Navbar/index.js]: Hum, that's subtle. I updated the code

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    window.addEventListener('resize', handleResize);
+    return () => window.removeEventListener('resize', handleResize);
+  }, []);
+
+  return windowSize;
+}
+
+export {desktopSize, mobileSize};
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js:42]: What about having an object like `WindowSizes` exposing mobile/desktop attributes

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+import {useEffect, useState} from 'react';
+
+const desktopThresholdWidth = 996;
+const desktopSize = 'desktop';
+const mobileSize = 'mobile';
+
+function useWindowSize() {
+  const isClient = typeof window === 'object';
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js:15]: It seems the build fails (check netlify deploy logs), I think you should use `typeof window !== "undefined"` maybe? not sure

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  function getSize() {
+    if (isClient) {
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js:18]: I think it's !isClient instead :)

Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js]: Indeed, my bad, went to quickly on that one.

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js (hunk 4 above)
Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js]: Looks like ```typeof window !== "undefined"``` is indeed more common.

File: packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js (hunk 3 above)
Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/hooks/useWindowSize.js]: Nice suggestion, I went and did that

File: packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js [hunk 6]
```diff
@@ -176,6 +177,13 @@ function DocSidebar(props) {
   } = props;
 
   useLockBodyScroll(showResponsiveSidebar);
+  const windowSize = useWindowSize();
+
+  useEffect(() => {
+    if (windowSize === windowSizes.desktop && showResponsiveSidebar) {
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/DocSidebar/index.js:183]: here too you could remove `showResponsiveSidebar ` (or you need to include it in useEffect deps)

File: packages/docusaurus-theme-classic/src/theme/Navbar/index.js [hunk 7]
```diff
@@ -201,6 +202,14 @@ function Navbar() {
     [setLightTheme, setDarkTheme],
   );
 
+  const windowSize = useWindowSize();
+
+  useEffect(() => {
+    if (windowSize === windowSizes.desktop) {
+      setSidebarShown(false);
+    }
+  }, [windowSize, sidebarShown]);
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Navbar/index.js:211]: now here sidebarShown is not used in the effect anymore so it's not necessary to add it to the deps :)

Comment (guillaumejacquart) [reply packages/docusaurus-theme-classic/src/theme/Navbar/index.js]: Indeed ! I made the changes

**Comment by slorber** [comment]
thanks :)
